Add start of ORM considerations (#2668)
dotnet/EntityFramework.Docs: entity-framework/core/index.md

---
title: Overview of Entity Framework Core - EF Core
description: General introductory overview of Entity Framework Core
author: ajcvickers
ms.date: 9/20/2020
uid: core/index
---

# Entity Framework Core

> Entity Framework (EF) Core is a lightweight, extensible, [open source](https://github.com/aspnet/EntityFrameworkCore) and cross-platform version of the popular Entity Framework data access technology.

Reviewer: wrong url!

Reviewer 2: It still redirects

EF Core can serve as an object-relational mapper (O/RM), which:

* Enables .NET developers to work with a database using .NET objects.
* Eliminates the need for most of the data-access code that typically needs to be written.

EF Core supports many database engines, see [Database Providers](xref:core/providers/index) for details.

## The model

With EF Core, data access is performed using a model. A model is made up of entity classes and a context object that represents a session with the database. The context object allows querying and saving data. For more information, see [Creating a Model](xref:core/modeling/index).

EF supports the following model development approaches:

* Generate a model from an existing database.
* Hand code a model to match the database.

> * Use [EF Migrations](xref:core/managing-schemas/migrations/index) to create a database from a model. Migrations allow evolving the database as the model changes.

Reviewer 3: This is a bit confusing to me, as Migrations aren't a way to develop/create your model, but rather to create (or sync) a database from a model... It may be worth improving this section and making that clearer ("Once your model has been created, you can use Migrations to keep your database schema in sync with it").

[!code-csharp[Main](../../samples/core/Intro/Model.cs)]

## Querying

Instances of your entity classes are retrieved from the database using Language [Integrated Query (LINQ)](/dotnet/csharp/programming-guide/concepts/linq/). For more information, see [Querying Data](xref:core/querying/index).

[!code-csharp[Main](../../samples/core/Intro/Program.cs#Querying)]

## Saving data

Data is created, deleted, and modified in the database using instances of your entity classes. See [Saving Data](xref:core/saving/index) to learn more.

[!code-csharp[Main](../../samples/core/Intro/Program.cs#SavingData)]

## EF ORM considerations

While EF Core is good at abstracting many programming details, there are some best practices applicable to any ORM that help to avoid common pitfalls in production apps:

 - Intermediate-level knowledge or higher of the underlying database server is essential to architect, debug, profile, and migrate data in high performance production apps. For example, knowledge of primary and foreign keys, constraints, indexes, normalization, DML and DDL statements, data types, profiling, etc.

> - Functional and integration testing:  It's important to replicate the production environment as close as possible to:

Reviewer 3: ```suggestion
- Functional and integration testing:  It's important to replicate the production environment as closely as possible to:
```

  - Find issues in the app that only show up when using a specific versions or edition of the database server .

>   - Catch breaking changes when upgrading EF Core and other dependencies. For example, adding or upgrading frameworks like ASP.NET Core, Application Insights, Serilog, or Automapper. These dependencies can affect EF Core in unexpected ways.

Reviewer 3: Definitely ASP.NET and Automapper (and e.g. OData), but App Insights and Serilog?

- Security review: For example, handling of connection strings and other secrets, database permissions for non-deployment operation, input validation for raw SQL, encryption for sensitive data.
- Make sure logging and diagnostics are sufficient and usable. For example, appropriate logging configuration, query tags, and Application Insights.
- Error recovery. Prepare contingencies for common failure scenarios such as version rollback, fallback servers, scale-out and load balancing, DoS mitigation, and data backups.
- Application deployment and migration. Plan out how migrations are going to be applied during deployment; doing it at application start can suffer from concurrency issues and requires higher permissions than necessary for normal operation. Use staging to facilitate recovery from fatal errors during migration. For more information, see [Applying Migrations](xref:core/managing-schemas/migrations/applying).

> - Detailed examination and testing of generated migrations. Migrations should be thoroughly tested before being applied to production data. The shape of the schema and the column types cannot be easily changed once the tables contain production data. For example, on SQL Server, `nvarchar(max)` and `decimal(18, 2)` are rarely the best types for columns mapped to string and decimal properties, but those are the defaults that EF uses because it doesn't have knowledge of your specific scenario.

Reviewer 3: If we're looking for why migrations should be inspected, then maybe EF generating DropColumn/AddColumn when you rename a property may be even more striking.

Out of curiosity, what's wrong with `decimal(18, 2)`? (am not a SQL Server guy)

Reviewer 2: If the client value has a larger scale then the extra digits will be truncated silently and there's no SQL type that can completely cover the CLR decimal range

Reviewer 3: Thanks @AndriySvyryd. So is it right to say that `decimal(18, 2)` is rarely the best type to map .NET decimal? I'm just thinking users are going to get alarmed here where a better option isn't necessarily available.

Reviewer 2: If you can phrase it in a way that indicates that the problem is in the length/precision/scale it would be better.

---
title: Overview of Entity Framework Core - EF Core
description: General introductory overview of Entity Framework Core
author: rowanmiller

> ms.date: 10/27/2016

Reviewer 2: Change author to ajcvickers and update date

Entity Framework (EF) Core is a lightweight, extensible, [open source](https://github.com/aspnet/EntityFrameworkCore) and cross-platform version of the popular Entity Framework data access technology.

EF Core can serve as an object-relational mapper (O/RM), which:

* Enables .NET developers to work with a database using .NET objects.
* Eliminates the need for most of the data-access code that typically needs to be written.

EF Core supports many database engines, see [Database Providers](xref:core/providers/index) for details.

> ## The Model

Reviewer 2: We should standardize to sentence case for non-title headings

Author: Per MS style guide, use sentence case.

With EF Core, data access is performed using a model. A model is made up of entity classes and a context object that represents a session with the database. The context object allows querying and saving data. For more information, see [Creating a Model](xref:core/modeling/index).

> EF supports the following for model development:

Reviewer 2: ```suggestion
EF supports the following model development approaches:
```

> * Generating a model from an existing database.

Reviewer 2: ```suggestion
* Generate a model from an existing database.
```

* Hand code a model to match the database.
* Use [EF Migrations](xref:core/managing-schemas/migrations/index) to create a database from a model. Migrations allow evolving the database as the model changes.

[!code-csharp[Main](../../samples/core/Intro/Model.cs)]

## Querying

> Instances of your entity classes are retrieved from the database using Language Integrated Query (LINQ). For more information, see [Querying Data](xref:core/querying/index).

Reviewer 2: Should we add a link to LINQ docs?

[!code-csharp[Main](../../samples/core/Intro/Program.cs#SavingData)]

## EF ORM considerations

> While EF Core is good at abstracting many programming details, there are some  aspects of EF Core that need to be handled for production apps to avoid common pitfalls:

Reviewer 2: ```suggestion
While EF Core is good at abstracting many programming details, there are some best practices applicable to any ORM that help to avoid common pitfalls in production apps:
```

 - Intermediate-level knowledge or higher of the underlying database server is essential to architect, debug, profile, and migrate data in high performance production apps. For example, knowledge of primary and foreign keys, constraints, indexes, normalization, DML and DDL statements, data types, profiling, etc.
- Functional and integration testing:  It's important to replicate the production environment as close as possible to:

>   - Find issues in the app.

Reviewer 2: ```suggestion
  - Find issues in the app that only show up when using a specific versions or edition of the database server .
```

  - Catch breaking changes when upgrading EF Core and other dependencies. For example, adding or upgrading frameworks like ASP.NET Core, Application Insights, Serilog, or Automapper. These dependencies can affect EF Core in unexpected ways.

> - Performance and stress testing with representative loads. The naive usage of some features doesn't scale well. For example, multiple collections Includes, heavy use of lazy loading, conditional queries on non-indexed columns, massive updates, and inserts with store-generated values, lack of concurrency handling, large models, inadequate cache policy.

Reviewer 2: ```suggestion
- Performance and stress testing with representative loads. The naive usage of some features doesn't scale well. For example, multiple collections Includes, heavy use of lazy loading, conditional queries on non-indexed columns, massive updates and inserts with store-generated values, lack of concurrency handling, large models, inadequate cache policy.
```

> - Security: For example, handling of connection strings, database permissions for non-deployment operation, input validation for raw SQL, encryption for sensitive data.

Reviewer 2: ```suggestion
- Security review: For example, handling of connection strings and other secrets, database permissions for non-deployment operation, input validation for raw SQL, encryption for sensitive data.
```

- Make sure logging and diagnostics are sufficient and usable. For example, appropriate logging configuration, query tags, and Application Insights.
- Error recovery. Prepare contingencies for common failure scenarios such as version rollback, fallback servers, scale-out and load balancing, DoS mitigation, and data backups.

> - Application deployment and migration. <!-- review I need some content to replace the following link. Links just provided as a handy ref. --> See https://github.com/dotnet/EntityFramework.Docs/issues/1879 and https://github.com/dotnet/EntityFramework.Docs/issues/814.

Reviewer 2: ```suggestion
- Application deployment and migration. Plan out how migrations are going to be applied during deployment; doing it at application start can suffer from concurrency issues and requires higher permissions than necessary for normal operation. Use staging to facilitate recovery from fatal errors during migration. For more information, see [Applying Migrations](xref:core/managing-schemas/migrations/applying).
```

> - Detailed examination and testing of generated migrations. Migrations should be thoroughly tested before being applied to production data. The shape of the schema and the column types cannot be easily changed once the tables contain production data. For example, on SqlServer, `nvarchar(max)` and `decimal(18, 2)` are rarely the best types for columns mapped to string and decimal properties.  <!-- review I need some content to replace the following link. Links just provided as a handy ref. --> See https://github.com/dotnet/efcore/issues/20159

Reviewer 2: ```suggestion
- Detailed examination and testing of generated migrations. Migrations should be thoroughly tested before being applied to production data. The shape of the schema and the column types cannot be easily changed once the tables contain production data. For example, on SQL Server, `nvarchar(max)` and `decimal(18, 2)` are rarely the best types for columns mapped to string and decimal properties, but those are the defaults that EF uses because it doesn't have knowledge of your specific scenario.
```
